### PR TITLE
Message filtering support

### PR DESCRIPTION
**Pull request by WyriHaximus**
As discussed in #11

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -27,7 +28,8 @@
  */
 class Plugin extends AbstractPlugin implements LoopAwareInterface
 {
-    const URL_HANDLER_INTERFACE = 'Phergie\Irc\Plugin\React\Url\UrlHandlerInterface';
+    const URL_HANDLER_INTERFACE  = 'Phergie\Irc\Plugin\React\Url\UrlHandlerInterface';
+    const EVENT_FILTER_INTERFACE = 'Phergie\Irc\Plugin\React\EventFilter\FilterInterface';
```
Comment (elazar) [inline src/Plugin.php]: Any particular reason for using a constant versus just referencing `FilterInterface::class` where needed?

File: src/Plugin.php [hunk 2]
```diff
@@ -113,6 +126,12 @@ public function logDebug($message)
 
     public function handleIrcReceived(UserEvent $event, EventQueue $queue)
     {
+        if (
+            $this->filter !== null &&
+            $this->filter->filter($event) !== false
+        ) {
+            return;
+        }
```
Comment (elazar) [inline src/Plugin.php]: The original impetus for filing #11 was to facilitate filtering out URLs by hostname. If the event filter doesn't have access to the parsed URLs or has to parse them a second time, it seems to defeat that point. It's already possible to use this plugin with the EventFilter plugin via configuration without any changes to this plugin (see [usage documentation](https://github.com/phergie/phergie-irc-plugin-react-eventfilter#usage)). What I want is filtering specific to URLs.

Comment (WyriHaximus) [inline src/Plugin.php]: You're right, currently writing a set of filters that lets user filter based on hostname, path, protocol, and port.

File: src/Plugin.php (hunk 1 above)
Comment (WyriHaximus) [inline src/Plugin.php]: Force of 5.4 habbit, I'll update both of them

**Comment by WyriHaximus** [comment]
Took care of the PHP 5.4 stuff and added a new filter that lets you filter on `parse_url` URL sections.

**Comment by WyriHaximus** [comment]
Rebased master into this PR so it can be merged without issues.

File: src/Filter/UrlSectionFilter.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    protected $value;
+
+    /**
+     * @var string
+     */
+    protected $section;
+
+    /**
+     *
+     * @param string $value
```
Comment (elazar) [inline src/Filter/UrlSectionFilter.php]: Need a `@param` tag for `$section`.

File: src/Filter/UrlEvent.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        return $this->event->getCommand();
+    }
+
+    public function getUrlSection($section)
+    {
+        if (isset($this->parsedUrl[$section])) {
+            return $this->parsedUrl[$section];
+        }
+
+        return false;
```
Comment (elazar) [inline src/Filter/UrlEvent.php]: Would returning `null` here would make more sense semantically (i.e. to imply the lack of a value)?

File: src/Filter/UrlSectionFilter.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public function filter(EventInterface $event)
+    {
+        if (!($event instanceof UrlEvent)) {
+            return null;
+        }
+
+        $section = $event->getUrlSection($this->section);
+        if ($section === false) {
+            return null;
```
Comment (elazar) [inline src/Filter/UrlSectionFilter.php]: This implies that the event should be allowed through if it doesn't contain the part of the URL that's being filtered on. Is that the desired behavior?

File: src/Filter/UrlEvent.php (hunk 4 above)
Comment (WyriHaximus) [inline src/Filter/UrlEvent.php]: The reasoning for using `false` is the failure of finding a value. If your preference is using `null` that can be arranged :smile:

File: src/Filter/UrlSectionFilter.php (hunk 5 above)
Comment (WyriHaximus) [inline src/Filter/UrlSectionFilter.php]: Yes if the part of the URL is missing it doesn't fall within scope of the filter. How ever in some cases it does make sense to return false if it is missing so I've added a strict bool in 2aed4c3
.

File: src/Filter/UrlEvent.php (hunk 4 above)
Comment (elazar) [inline src/Filter/UrlEvent.php]: Yeah, I think `null` makes more sense here.

Comment (WyriHaximus) [inline src/Filter/UrlEvent.php]: Done

**Comment by elazar** [comment]
:+1:

**Comment by WyriHaximus** [comment]
@svpernova09 @PSchwisow Any of you has any comments on this? Otherwise I'de like to merge and tag 3.1.0 tomorrow.

**Comment by svpernova09** [comment]
:+1: thanks! Looks good from here

**Comment by WyriHaximus** [comment]
Will let travis run it's course before tagging to be sure nothing funny pops up

**Pull request merged**